cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/internal/documentstore/DatabaseImpl.java

```diff
@@ -1191,4 +1186,43 @@ public DocumentRevision delete(final DocumentRevision rev) throws
         }
     }
 
+    @Override
+    public void createWithHistory(DocumentRevision revision, int revisionsStart, List<String>
+            revisionsIDs) throws DocumentException {
+
+        Map<String, Attachment> attachments = revision.getAttachments();
+        Collections.list(Collections.enumeration(revision.getAttachments().values()));
+
+        InternalDocumentRevision internalRev = new DocumentRevisionBuilder()
+                .setDocId(revision.getId())
+                .setRevId(revision.getRevision())
+                .setBody(revision.getBody())
+                .setAttachments(attachments)
+                .setDeleted(revision.isDeleted())
+                .build();
+
+        Map<String[], Map<String, PreparedAttachment>> preparedAttachments = Collections
+                .singletonMap
+                        (new String[]{revision.getId(), revision.getRevision()}, AttachmentManager
+                                .prepareAttachments(attachmentsDir, attachmentStreamFactory,
+                                        attachments));
+
+        Map<String, Object> attachmentsMetadata = new HashMap<String, Object>();
```

Comment:
Since the caller has to initialize the `DocumentRevision` that is passed to `createWithHistory` the idea was that the attachment metadata setting that you describe would be performed by the caller since we don't do any direct JSON parsing as part of this.

I tried to indicate this intent in the [javadoc example](https://github.com/cloudant/sync-android/blob/14ed6536a7312e9426f9330f2bd355a9ae755349/cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/documentstore/advanced/XDatabase.java#L60)

Perhaps that needs to be elaborated further.